text/0000-inline-const.md

```diff
@@ -0,0 +1,276 @@
+named `const`), so the following are all guaranteed to work:
+
+```rust
+let x: &'static i32 = &const { 4i32.pow(4) }; // NOT IDIOMATIC
+let x: &'static i32 = const { &4i32.pow(4) }; // IDIOMATIC
+
+// If RFC 2203 is stabilized
+let v = [const { Vec::new() }; 3];  // IDIOMATIC
+let v = const { [ Vec::new(); 3] }; // NOT IDIOMATIC
+```
+
+Whether to lint against the non idiomatic versions is an open question.
+Personally, I would like to lint against `&const {...}` but not `const { [expr;
+3] }`.
+
+Inline `const`s are allowed within `const` and `static` initializers, just as we
+currently allow nested `const` declarations. Whether to lint against inline
+`const` expressions inside a `const` or `static` is also an open question.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This excludes other uses of the `const` keyword in expressions and patterns.
+I'm not aware of any other proposals that would take advantage of this.
+
+This would also be the first use of type inference for const initializers. I'm
+not aware of any technical issues that would arise from this, but perhaps I'm
+overlooking something?
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The main alternative is the status quo. Maintaining it will likely result in
+promotion being used for more contexts. The lang-team decided to [explore this
+approach](https://github.com/rust-lang/rust/pull/70042#issuecomment-612221597)
+instead.
+
+It would also possible to separate out the parts of this RFC relating to patterns
+so that they can be decided upon seperately. I think they are similar enough
+that they are best considered as a unit, however.
+
+# Prior art
+[prior-art]: #prior-art
+
+I'm not aware of equivalents in other languages.
+
+AFAIK, this was [first proposed] by @scottmcm.
+
+[first proposed]: https://internals.rust-lang.org/t/quick-thought-const-blocks/7803/9
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## Naming
+
+I prefer the name inline `const`, since it signals that there is no difference
+between a named `const` and an inline one. @scottmcm prefers "`const` block",
+which is closer to the syntax and parallels the current terminology of `async`
+block and `unsafe` block. It also avoids any accidental conflation with the
+`#[inline]` attribute, which is unrelated.
+
+## Lints
+
+As mentioned in the reference-level specification, we need to decide whether we
+want to lint against certain types of inline `const` expressions.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+It would be possible to allow the syntax `const expr` for an inline `const` that
+consists of a single expression. This is analagous to the single expression
+variant of closures: `|| 42`. This is backwards compatible with the current proposal.
+
+This could allow us to deprecate the more esoteric classes of promotable
+expressions (e.g., `&(u32::MAX + u32::MAX)`) in favor of inline `const`
```

Comment:
Is this esoteric only because it gives a `const_err` message?
What does this proposal do in this case?

Reply:
This does not give a `const_err` message currently. It does trigger the `arithmetic_overflow` lint, however. Writing `const {u32::MAX + u32::MAX }` would indeed trigger the `const_err` lint. This proposal makes no change to the rules for implicit promotion.

Reply:
> This does not result in a `const_err` lint currently

Yes it does: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6ef001057822f9a1f070e973e5c3ff0f

But that still doesn't answer my first question: What makes this esoteric? What's "the more esoteric classes of promotable expressions" this sentence talks about?

Reply:
That's not promotion. That's an explicit const context. [This is promotion](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=67fa7a9353105190cd32dbc7f99681d0). Perhaps the fact that we're having this conversation indicates that the promotion of arithmetic expressions is indeed esoteric.

Reply:
@ecstatic-morse so the point is, you want to not only deprecate calling functions in implicit promotion contexts (the few that are flagged `#[rustc_promotable]`), but also arithmetic? Given that arithmetic can panic, I tend to agree this is rather subtle. But please say so more explicitly -- I don't think "esoteric" conveys much meaning here.

Reply:
You're right. I could have been more explicit.

I think that, if this RFC is merged and, separately, we stop promoting some expressions, we should just do them all. I don't really wanna have to remember that unary negation can panic but bitwise negation cannot, for example. Also, it's worth noting that bitwise shifts can fail.

Reply:
> I don't really wanna have to remember that arithmetic negation can overflow but bitwise negation cannot, for example. Also, it's worth noting that bitwise shifts can fail.

Fair.
The things that are left are basically all also allowed in patterns, right? That's a nice symmetry.